### ISSUE TITLE
Cannot read property 'constructor' of undefined 

### ISSUE THREAD

**Issue by eanderton** [gopherjs/gopherjs#237]
I ran into this one when trying to use Go functions that are callable from JavaScript code.

Playground link:
http://www.gopherjs.org/playground/#/A_bP7IWgxi

Basically, $internalize has no code path for 'undefined' values that are mapped to interfaces.  

I would like to recommend the following fix, for line 245 in compiler/prelude/jsmapping.go like so:
```
    if (v === null || v === undefined) {
      return $ifaceNil;
    }
```
This changes $internalize such that undefined is converted to nil, just like null, for interfaces.

**Comment by shurcooL** [comment]
That looks like a reasonable fix. But have you considered that there's a special object for representing `undefined` in GopherJS?

See http://godoc.org/github.com/gopherjs/gopherjs/js#Undefined.

Perhaps a better fix would be making it so that `func foo(x interface{})` is called with `x` being `*js.Undefined` instead of `nil`?

That way there's a bijection between Go `nil` and JavaScript `null`, and Go `*js.Undefined` and JavaScript `undefined`.

**Comment by eanderton** [comment]
@shurcooL - That would be even better.  It fits with the rest of GopherJS' design.

**Comment by Archs** [comment]
@shurcooL That's Great  :+1:

**Comment by shurcooL** [comment]
I'm working on this now, I think I'll have a PR that fixes this for @neelance to review soon. It looks very nice and consistent so far. I'm going to convert my manual test into an automatic one and make that PR.

**Comment by shurcooL** [comment]
I've started my fix at #240.

I wrote the following short program while testing my work:

https://play.golang.org/p/49RFm44M9O (it won't run on Go Playground because it needs gopherjs.com/gopherjs/gopherjs/js import)

I would've made it into an example test, but `gopherjs test` doesn't support that at this time (see #239), so I converted it into a normal test, which made it more verbose unfortunately. See https://github.com/shurcooL/gopherjs/commit/2f0c1d0a280ed8e6f74ea4bea0126be834c1b56b.

The test should be failing now. Next up I'll add the commit that fixes the issue (and hopefully tests pass).
